ppl/zqd/space/compactor.go

```diff
@@ -0,0 +1,120 @@
+package space
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/brimsec/zq/api"
+	"github.com/brimsec/zq/ppl/zqd/storage/archivestore"
+	"go.uber.org/zap"
+	"golang.org/x/sync/semaphore"
+)
+
+type compactor struct {
+	cancel      context.CancelFunc
+	done        sync.WaitGroup
+	logger      *zap.Logger
+	manager     *Manager
+	notify      chan api.SpaceID
+	compactDone chan api.SpaceID
+	sem         *semaphore.Weighted
+}
+
+const maxConcurrentCompacts = 1
+
+func newCompactor(manager *Manager) *compactor {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &compactor{
+		cancel:      cancel,
+		logger:      manager.logger.Named("compactor"),
+		manager:     manager,
+		notify:      make(chan api.SpaceID),
+		compactDone: make(chan api.SpaceID),
+		sem:         semaphore.NewWeighted(maxConcurrentCompacts),
+	}
+	go c.run(ctx)
+	return c
+}
+
+func (c *compactor) WriteComplete(id api.SpaceID) {
+	c.notify <- id
+}
+
+func (c *compactor) launchCompact(ctx context.Context, id api.SpaceID) {
+	go func() {
+		if err := c.sem.Acquire(ctx, 1); err != nil {
+			return
+		}
+		defer c.sem.Release(1)
+		defer func() { c.compactDone <- id }()
+		c.compact(ctx, id)
+	}()
+}
+
+func (c *compactor) run(ctx context.Context) {
+	c.done.Add(1)
```

Comment:
This should happen before the goroutine starts. I suggest moving it and the Done call into newCompactor like this:
```
	c.done.Add(1)
	go func() {
		c.run()
		c.done.Done()
	}